Fix `always()` logic for dependent tasks in CI/CD workflow

AetherUnbound opened the pull request:
## Description

<!-- Concisely describe what the pull request does. -->
<!-- Add screenshots, videos, or other media to show the problem and the solution when appropriate. -->

I noticed that we weren't receiving any API deployment messages in Slack, even though we've merged PRs which _should_ have initiated an API deploy. After some investigation, I found that we were requiring changes made to the ingestion server prior to deploying the API:

https://github.com/WordPress/openverse/blob/a2782e8b45300be669c90b4da1abd801caa0263b/.github/workflows/ci_cd.yml#L926

Since these services are distinct, we shouldn't need changes made to the ingestion server in order to deploy the API. This PR removes that requirement.

## Testing Instructions

<!-- Give steps for the reviewer to verify that this PR fixes the problem; or delete this section entirely. -->

Merge and hope :slightly_smiling_face:

## Checklist

<!-- Replace  the [ ] with [x] to check the boxes. -->

- [x] My pull request has a descriptive title (not a vague title like
      `Update index.md`).
- [x] My pull request targets the _default_ branch of the repository (`main`) or
      a parent feature branch.
- [x] My commit messages follow [best practices][best_practices].
- [x] My code follows the established code style of the repository.
- [ ] I added or updated tests for the changes I made (if applicable).
- [ ] I added or updated documentation (if applicable).
- [ ] I tried running the project locally and verified that there are no visible
      errors.

[best_practices]:
  https://git-scm.com/book/en/v2/Distributed-Git-Contributing-to-a-Project#_commit_guidelines

## Developer Certificate of Origin

<!-- You must read and understand the following attestation. -->

<details>
<summary>Developer Certificate of Origin</summary>

```
Developer Certificate of Origin
Version 1.1

Copyright (C) 2004, 2006 The Linux Foundation and its contributors.
1 Letterman Drive
Suite D4700
San Francisco, CA, 94129

Everyone is permitted to copy and distribute verbatim copies of this
license document, but changing it is not allowed.


Developer's Certificate of Origin 1.1

By making a contribution to this project, I certify that:

(a) The contribution was created in whole or in part by me and I
    have the right to submit it under the open source license
    indicated in the file; or

(b) The contribution is based upon previous work that, to the best
    of my knowledge, is covered under an appropriate open source
    license and I have the right under that license to submit that
    work with modifications, whether created in whole or in part
    by me, under the same open source license (unless I am
    permitted to submit under a different license), as indicated
    in the file; or

(c) The contribution was provided directly to me by some other
    person who certified (a), (b) or (c) and I have not modified
    it.

(d) I understand and agree that this project and the contribution
    are public and that a record of the contribution (including all
    personal information I submit with it, including my sign-off) is
    maintained indefinitely and may be redistributed consistent with
    this project or the open source license(s) involved.
```

</details>

obulat reviewed (commented):
I don't understand why we should need this change. I read the old code as "If anything in the API **or** the ingestion server changes, deploy". So, it **should** deploy if there are only changes in the API. 
With the changes here, will deploy not happen when the ingestion server is changed? Don't we want that?

obulat commented:
I compared two runs (after the dependabot PRs were merged). [One](https://github.com/WordPress/openverse/actions/runs/4309734336) had changes to the API, and the [other](https://github.com/WordPress/openverse/actions/runs/4310292072) - to ingestion server.

The ingestion server one did deploy. I think the difference is in the "Publish Docker images": it does not run if the ingestion server tests didn't run.
This `needs` step is not run, and the API is not deployed if changes are in API only:
https://github.com/WordPress/openverse/blob/1621ed0d68daf62420f997c97745f361191a9825/.github/workflows/ci_cd.yml#L931

AetherUnbound commented:
> Sorry, I was comparing the wrong runs: the "Publish Docker images" only runs on the "push" event, and it does run when there are api changes: https://github.com/WordPress/openverse/actions/runs/4309530451

This run is a good example though - the ingestion server tests didn't run and so a staging deployment _of the API_ wasn't run, even though it should be. At least, that's what I suspect is happening 🙂 GitHub Actions doesn't give you much info beyond "this was skipped" so it's kind of hard to tell haha.

dhruvkb commented:
@obulat's understanding is correct, the deploy happens if either API or ingestion server has changed.

We will need to use the expanded form of the `if` condition which is seen in some other jobs.

dhruvkb commented:
<img width="1029" alt="Screenshot 2023-03-02 at 10 12 50 AM" src="https://user-images.githubusercontent.com/16580576/222345986-dd340ca6-d52e-4c10-8955-2186d5d907f5.png">

What I don't get is why `deploy-api` was skipped even though all  its dependencies have passed.

AetherUnbound commented:
Ah okay, thank you both. I was misunderstanding the situation and the code I altered. That said, I think the particular line change I made might still apply, since we don't have ingestion server deployments set up automatically (so no need to deploy the API on ingestion server changes). What do y'all think? I'm also confused about that @dhruvkb, I'll try to look into it more today.

AetherUnbound commented:
Here are three example runs. All 3 affect the API, but the first 2 do not initiate API deployments. The only difference I can see from the 3rd (which _did_ have a successful run) was the fact that the ingestion server tests were not skipped in this case:
- https://github.com/WordPress/openverse/actions/runs/4309189828
- https://github.com/WordPress/openverse/actions/runs/4314531114
- https://github.com/WordPress/openverse/actions/runs/4310292072

AetherUnbound commented:
@dhruvkb I think this _has_ to be an interaction with the `always() &&` logic and these lines in the `publish-images` task:
https://github.com/WordPress/openverse/blob/1621ed0d68daf62420f997c97745f361191a9825/.github/workflows/ci_cd.yml#L854-L855

I suspect this gets us into some weird behavior where the job itself runs but due to its dependencies being skipped it registers as "skipped" for downstream tasks.

AetherUnbound commented:
Ha, this behavior is exactly described in the next comment on the original suggestion 😅 🤦🏼‍♀️ https://github.com/actions/runner/issues/491#issuecomment-882886236

AetherUnbound commented:
OKAY I think I got it! Per some of the discussion in that linked thread, I've changed may of our `always()` calls to something more appropriate, since `always()` will run even if the job is cancelled or fails. I also added that condition to the API deploy step so that it ignores the skipped state of its grandparents and runs when it should. Fingers crossed that this does it! 🤞🏼 🍀

dhruvkb commented on .github/workflows/ci_cd.yml:
```diff
@@ -923,7 +926,10 @@ jobs:
     name: Deploy staging API
     runs-on: ubuntu-latest
-    if: github.event_name == 'push' && (needs.get-changes.outputs.ingestion_server == 'true' || needs.get-changes.outputs.api == 'true')
+    if: |
+      !failure() && !cancelled() &&
+      github.event_name == 'push' && needs.get-changes.outputs.api == 'true' &&
+      (needs.test-ing.result == 'success' && needs.publish-images.result == 'success')
```
A similar change could be needed for the `deploy-frontend` action above which also depends on the sometimes-skipped `nuxt-playwright-e2e`, `nuxt-playwright-vr` and `storybook-playwright` jobs.

AetherUnbound replied:
Ah, thanks for that callout!

AetherUnbound replied:
I've taken a stab at this here: 8bb3fef7823b7512ce11690b512cdd790f8cae31 - do you think that looks right?

dhruvkb replied:
That looks logically right 👍. You could break it into several lines for readability but nonetheless, it should work just fine.

dhruvkb reviewed (approved):
LGTM, except for a request to update the frontend `if` condition as well. I won't block it though.

It's unfortunate that the only way to check if these changes work is push and pray.

obulat reviewed (approved):
It's unfortunate that `always()` acts in such an nonintuitive way in this case. Thank you for sharing the steps of your investigation!
🤞
